pkg/utils/nodes.go

```diff
@@ -0,0 +1,24 @@
+package utils
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// isNodeNameValid returns an error if nodeName doesn't match any node name int the cluster, otherwise a nil
+func IsNodeNameValid(r client.Reader, nodeName string) error {
+	node := &corev1.Node{}
+	objNodeName := types.NamespacedName{Name: nodeName}
+	err := r.Get(context.Background(), objNodeName, node)
+	if err != nil {
+		// Build node not found error
+		msg := fmt.Sprintf("Failed: node %s is not part of the cluster's node names", nodeName)
+		err = buildFailureApiMessage(err, msg)
+		return fmt.Errorf("CR's name doesn't match any node name in the cluster - %w", err)
```

Comment:
1. Logging is weird:

- 2 lines above: "node %s is not part of the cluster's node names" -> not true, you just failed to get it for SOME reason
- here: "CR's name doesn't match any node name in the cluster" -> yes, but maybe just because you failed to get the node for a reason which isn't notFound

2. Don't you need to differentiate error handling based on the error type? I think you have 3 cases here:
- node found  = name valid
- node not found = name invalid
- any other error = retry

Does that make sense?

With that in mind, the func signature might need an update as well (atm the `is` prefix suggests that the return value is a boolean, bit it's an error)

Reply:
I agree on 1and I hope now it is better.

> Don't you need to differentiate error handling based on the error type?
I am trying to do so using `buildApiError` function

>any other error = retry

I am not fully understand why to retry getting the node?

Reply:
> > Don't you need to differentiate error handling based on the error type?

>  I am trying to do so using `buildApiError` function
 
I don't see that? 🤔 You just return another error, no matter which error occurred in first place

> > any other error = retry
 
> I am not fully understand why to retry getting the node?

Because NotFound is the only error which says: there is no node with that name.
Every other error is: I had an issue while getting the node.

Reply:
> You just return another error, no matter which error occurred in first place

I create new error with apiErrors based on the given error type (see the use of [IsNotFound](https://github.com/medik8s/fence-agents-remediation/pull/48/files#diff-cf316a9ab7edc81ea1ea26252702d68f8286e9693eb9a284073535a9b0e21301R15), and [IsForbidden](https://github.com/medik8s/fence-agents-remediation/pull/48/files#diff-cf316a9ab7edc81ea1ea26252702d68f8286e9693eb9a284073535a9b0e21301R23) functions). 
There might be a better way of doing that by updating the current error, but the **_[buildApiError](https://github.com/medik8s/fence-agents-remediation/pull/48/files#diff-cf316a9ab7edc81ea1ea26252702d68f8286e9693eb9a284073535a9b0e21301R12)_** function does consider the input error type.

Reply:
> Every other error is: I had an issue while getting the node.

Agreed, I just don't understand why we need to retry getting the node. It sounds like you are suggesting to perform a recursion of calling CheckNodeName if it gets an error, when it tries to find a node with a matching `nodeName`.

ATM if [err](https://github.com/medik8s/fence-agents-remediation/pull/48/files#diff-44072254fdc9157bd09396bfdafca285a425236352804c1252445a87e47716ebR17) is not of type _notFound_ and not _forbidden_, then the buildApiError will use the [default error](https://github.com/medik8s/fence-agents-remediation/pull/48/files#diff-cf316a9ab7edc81ea1ea26252702d68f8286e9693eb9a284073535a9b0e21301R33) for the "retry" case you presented.

Reply:
> I just don't understand why we need to retry getting the node.

Because you don't know if the node exists or not. So you don't know if the CR is valid or not.

> It sounds like you are suggesting to perform a recursion of calling CheckNodeName if it gets an error,

No, not recursion. Stop reconcile by returning the error, which results in requeue with backoff.

Reply:
> Does this mean that, in Reconcile, you return with an error either way?

Correct

> Wouldn't be okay to requeue, instead?

> Stop reconcile by returning the error, which results in requeue with backoff.

IIUC both of you are suggesting to requeue with a backoff (e.g. 10 seconds) when the error is not of type _notFound_ and not _forbidden_. I added it now.
But I still wonder why waiting a backoff would be better. It sounds like you suspect that the `notFound` error is temporary, and after the backoff it will be resolved. 
Could be, but if not than we should add in reconcile a flag/counter to stop future backoffs.
WDYT?

Reply:
> But I still wonder why waiting a backoff would be better. 

No hardcoded backoff, just return an error. `controller-runtime` will requeue with an increasing backoff automatically. This ensures to retry but not overload a potentailly already broken api server.